### PR TITLE
Angelou

### LINKED ISSUE #92: Sign Up for Frost Account
As a Po.et user, I want to be able to Sign Up with Frost so that I can register and view my works in the blockchain.

* [x] Sign Up Form
* [x] Verify Password
* [x] Register with Frost-Client
* [x] Redirect to "Dashboard"

### PR DESCRIPTION
**Pull request by jwicks31**
Resolves #92, Part of #93 

### Product Spec

-As a Po.et user, I want to be able to Sign Up with Frost so that I can register and view my works in the blockchain.
- Log In/ Sign Up buttons in NavBar on landing page 
- Update Footer

### Features Toggles 
- '?ft=nav-buttons' - Show login/sign up buttons on landing page,
- '?ft=sign-up,auth' - Shows Sign Up Form when visiting `/register` + allows for registration,
- '?ft=footer', - Shows additional links - pages not yet built

### PR Review Checklist

* [ ] **Do the changes match the product spec?**
* [ ] **Are there changes in this PR that weren't originally in the product spec?**
* [ ] **Is a feature toggle introduced for this PR? Is it disabled by default?**
* [ ] **Have you tested the changes?**
* [ ] **Is QE review required? If so, has this been reviewed by QE?** (For all new features, review by QE is required. Minor changes which already have passing functional tests may not need a new QE review -- when in doubt, loop in a QE reviewer).
* [ ] **Are there code changes that require new unit tests? If so, are those tests included?** (A separate PR for tests is not appropriate).
* [ ] **Are documentation updates required? If so, does this PR include it?** (Separate documentation PR is generally not appropriate).
* [ ] **Are the added/edited files named appropriately and located correctly in the project directory structure?**

### Warning

Merging to `master` automatically triggers a deploy in production.

### CODE REVIEW

**Comment by lautarodragan** [comment]
@jwicks31 is this one ready for review?

**Comment by jwicks31** [comment]
@lautarodragan Yup!

**Comment by lautarodragan** [comment]
Manual testing:

Tried signing up with 
- a password that doesn't have required complexity,
- non-matching passwords,

These cases were rejected and notified as expected. So far so good.

Then tried signing up with a proper password. Seemed successful, but I was redirected to the home page and wasn't notified of anything at all.

In case this was due to `ft` query param being lost in the redirect I tried manually adding `/?ft=nav-buttons,footer,register,auth,login` but I still see the navbar showing the `Sign Up` and `Log In` buttons.

I did receive the `Po.et Verify Account` mail, though, so the sign up seems to have been done correctly.

After this, if I navigate to /login?ft=auth,login I get automatically redirected to the home page, which would indicate that I'm actually signed in. Looking at the local storage seems to confirm this: `reduxPersism:user` has the data of my just-created account.

So it would seem it boils down to two UX issues:
- No proper notification to the user after successful sign up
- NavBar keeps showing Log In and Sign Up buttons while signed in

I'm guessing the second is addressed in https://github.com/poetapp/explorer-web/pull/166. Can you confirm the first one @jwicks31?

All else, from the manual testing point of view, seems good. Still need to finish going over the code diff.

File: src/components/atoms/InputPassword/InputPassword.tsx
Comment (lautarodragan) [inline src/components/atoms/InputPassword/InputPassword.tsx:1]: Is this copied over from Frost?

File: src/components/containers/SignUp.container.tsx
Comment (lautarodragan) [inline src/components/containers/SignUp.container.tsx]: Isn't this method just boilerplate? Why not simply `return <SignUpWrapper onSubmitSignUp={this.props.onSubmitSignUp} signUp={signUp} />` in the render?

File: src/components/molecules/Forms/SignUp/SignUp.tsx
Comment (lautarodragan) [inline src/components/molecules/Forms/SignUp/SignUp.tsx]: Just curious — can this be typed as a React Synthetic Event rather than any?

File: src/pages/SignUp/SignUp.loader.tsx
Comment (lautarodragan) [inline src/pages/SignUp/SignUp.loader.tsx]: Why not make the route `sign-up` for consistency? (components and other stuff seem to be named `sign-up`)

Also: I find `sign up` matches more nicely to `sign in` and `register` does to `log in`. But that's just my personal preference.

File: src/sagas/SignUp.ts
Comment (lautarodragan) [inline src/sagas/SignUp.ts]: Now that https://github.com/poetapp/explorer-web/pull/147 is merged, can you follow the same pattern?

File: src/reducers/User.ts
Comment (lautarodragan) [inline src/reducers/User.ts:25]: This looks like a great place for bugs to hide. I see this pattern was already in use before this PR, so I created [this ticket](https://github.com/poetapp/explorer-web/issues/181) which we can address separately, not to hold this PR that much.

**Review by lautarodragan — CHANGES_REQUESTED** [review]
See comments.

File: src/components/atoms/InputPassword/InputPassword.tsx
Comment (jwicks31) [reply src/components/atoms/InputPassword/InputPassword.tsx:1]: Yes

**Comment by jwicks31** [comment]
@lautarodragan 
- NavBar keeps showing Log In and Sign Up buttons while signed in

It won't technically be fixed in that PR, since the sign up buttons aren't yet included in that branch, but it would be easier to implement with the updated Navbar component that is introduced in that #166/ #169. 

- No proper notification to the user after successful sign up

This will eventually take you to a profile page, which I think would convey that the sign up happened successfully. Do you think we will need more of a notification?

File: src/components/molecules/Form/Form.stories.tsx
Comment (wzalazar) [inline src/components/molecules/Form/Form.stories.tsx]: you should import it before to relative import, what you say? (cosmetic)

File: src/components/molecules/Forms/SignUp/SignUp.tsx
Comment (wzalazar) [inline src/components/molecules/Forms/SignUp/SignUp.tsx]: maybe we can share the logic 

```
 if (password === confirmPassword || value === '') elements.confirmPassword.setCustomValidity('')
```
instead of:

```
   if (password === confirmPassword) elements.confirmPassword.setCustomValidity('')
   if (value === '') elements.confirmPassword.setCustomValidity('')
```

what do you say?

File: src/components/containers/SignUp.container.tsx
Comment (jwicks31) [reply src/components/containers/SignUp.container.tsx]: Yeah. What do you think about getting rid of the wrapper completely and doing: 
```tsx
export const SignUpContainer = connect(mapStateToProps, mapDispatchToProps)(
  ({ signUp: { loading, error }, onSubmitSignUp }: SignUpContainerProps) => (
    <SignUp onSubmit={onSubmitSignUp} disabledButton={loading} serverErrors={error} />
  )
)
```

File: src/components/molecules/Forms/SignUp/SignUp.tsx
Comment (jwicks31) [reply src/components/molecules/Forms/SignUp/SignUp.tsx]: It lets me if I cast the target type afterward:

```tsx
readonly onChangeEmail = (e: React.SyntheticEvent): void => {
    const input = e.target as HTMLInputElement;
    input.setCustomValidity('')
  }
```
and 
```tsx
readonly onChangeRepeatPassword = (e: React.SyntheticEvent, data: any, elements: any): void => {
    const target = e.target as HTMLInputElement;
    const value = target.value;
```

File: src/components/molecules/Forms/SignUp/SignUp.tsx
Comment (jwicks31) [reply src/components/molecules/Forms/SignUp/SignUp.tsx]: LGTM!

File: src/pages/SignUp/SignUp.loader.tsx
Comment (jwicks31) [reply src/pages/SignUp/SignUp.loader.tsx]: I was going back and forth on updating everything to either be register or sign-up. I'll change this to sign-up for now. As far as the pairing I'm not sure that I have a real preference there, but in the Invision design they use `Sign Up` / `Log In`

File: src/components/containers/SignUp.container.tsx
Comment (lautarodragan) [reply src/components/containers/SignUp.container.tsx]: LGTM!

File: src/components/molecules/Forms/SignUp/SignUp.tsx
Comment (lautarodragan) [reply src/components/molecules/Forms/SignUp/SignUp.tsx]: That looks good!

Comment (lautarodragan) [reply src/components/molecules/Forms/SignUp/SignUp.tsx]: OK, I went and checked an old project of mine and then remembered the correct way to do this.

`SyntheticEvent` takes a generic parameter than can be used to type `currentTarget`. 

I think it would look like this:

```ts
readonly onChangeRepeatPassword = (e: React.SyntheticEvent<HTMLInputElement>, data: any, elements: any): void => {
    const target = e.currentTarget;
    const value = target.value;
```

There's also many more-specific interfaces that extend from SyntheticEvent, such as `KeyboardEvent` and `MouseEvent`.

Here's the [source code of the React typings](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L567).

Here's a [`target` vs `currentTarget`](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12239) related discussion. 

And here's [documentation on `currentTarget`](https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget).

For our current use case I don't think we really need to change anything in this PR, but I wanted to leave this here as some sort of documentation on the topic.

I believe in onChange events using target and currentTarget has the same effect, and using the generic parameter of SyntheticEvent or casting doesn't make any real different either (since neither do any actual type checking).

@kennylavender @jwicks31 @wzalazar

**Review by lautarodragan — APPROVED** [review]

**Review by lautarodragan — DISMISSED** [review]
@wzalazar can you re-review please?

File: src/pages/Login/Login.layout.tsx
Comment (wzalazar) [inline src/pages/Login/Login.layout.tsx]: absolute path?

File: src/sagas/SignUp.test.ts
Comment (wzalazar) [inline src/sagas/SignUp.test.ts]: can you add an empty line before `assert`?

Comment (jwicks31) [reply src/sagas/SignUp.test.ts]: Done!

**Comment by jwicks31** [comment]
@lautarodragan My last commits dismissed your review

**Pull request merged**
